### PR TITLE
wip: Fixes for user-status model

### PR DESCRIPTION
**Pull request by chrisbobbe**
This has a lot of cleanups, followed by a `wip` commit for more intentionally aligning with servers' behavior, based on my observations at https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/Emoji.20statuses.20in.20zulip.2Eyaml/near/1322329

I figured it's time to let @gnprice have a look at this work, since we'd like to get #4925 done along with the 5.0 release soon. 🙂

Still to do:
- Get those observations into the doc if they're right
- More tests for the updated model (should go in that commit at the end, or a separate commit after?)
- Consolidate reducer/selector files into a `userStatusModel` file
- Rebase atop #5223 once merged (use new `mkActionRegisterComplete`, etc.)
- UI for emoji statuses

### CODE REVIEW

File: src/api/modelTypes.js [hunk 1]
```diff
@@ -211,28 +211,6 @@ export type UserGroup = $ReadOnly<{|
...
 
-/**
- * Specifies user status related properties
- * @prop away - present if we are to override user's presence status
- *   * This is the "user status" / "unavailable" feature added in early 2019.
- *     (At time of writing, there are no docs to link to.)
- * @prop status_text - a string representing information the user decided to
- *   manually set as their 'current status'
- */
-export type UserStatus = $ReadOnly<{|
```
Comment (gnprice) [inline src/api/modelTypes.js:222]: Dang, too bad this doesn't turn out to quite align between the initial data and the event.

Let's have "see also" comments pointing bidirectionally between those, though, to help the reader at least see what to compare to.

File: src/api/initialDataTypes.js [hunk 2]
```diff
@@ -499,13 +498,47 @@ export type InitialDataUpdateMessageFlags = $ReadOnly<{|
...
+/**
+ * Current values for (all?) users' self-assigned availability and
+ *   text/emoji statuses
+ */
+// TODO: Find out if servers ignore any users (e.g., deactivated ones) when
+//   assembling this.
 export type InitialDataUserStatus = $ReadOnly<{|
   /**
    * Older servers (through at least 1.9.1) don't send this.
-   * A missing value is equivalent to empty.
```
Comment (gnprice) [inline src/api/initialDataTypes.js:505]: types [nfc]: Remove misleading bit of jsdoc on initial-data `user_status`
    
    Looks like this was added in c8d071452.
    
    I think it's unlikely that `user_status` has ever been missing
    except because the app is talking to an old server that doesn't
    support `user_status`.

Yeah, I believe what was meant by this line was: if this is missing, that means this is an old server which doesn't have the `user_status` feature, and then the meaning of that is the same as if we got a `user_status` value that was empty.


    A missing `user_status` from an old server doesn't mean the same
    thing as a present but empty `user_status` from a current server.
    
    In particular:
    
    - a missing `user_status` from an old server means we just don't
      know what users' statuses are.

Hmm. I don't think I quite agree with this interpretation.

If the server doesn't have the feature of user status, then I would say that there is no such thing as the status of a user on that server. It's not that they really do have a status of ":calendar: In a meeting" or ":palm_tree: Vacationing" or whatever and we don't know what it is -- we do know the status of each user, and its value is… no-status.

It's true that out in the real world, the human behind that user may really be in a meeting, or vacationing, or so on, and we don't know which of those they are. But it's not like we know that real-world state, or even have all that solid a guess at it, when the feature does exist. (People tend to set these things and not update them. Or interpreted another way, they use them abstractly as signifiers -- they leave it at "In a meeting", say, not because they're literally always in meetings, nor by carelessness, but because they want to discourage you from PMing them and expecting a quick response.) So the "user status" we're concerned with isn't that real-world state of the human in any case -- rather it's whatever value they've chosen to communicate through this product feature. Which, when the feature doesn't exist, is no value.

Now the question is, how should we handle having no status value?

    - a present but empty `user_status` from a current server means
      everyone's status is boring (in particular, everyone is
      "available" and no one currently has a text or emoji status set);
      see
        https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/Emoji.20statuses.20in.20zulip.2Eyaml/near/1322329 .

I don't think it's right to say that a boring status, on a modern server where user status exists, means the person is "available". Here's the user-facing doc:
https://zulip.com/help/status-and-availability

> Your **availability** is a colored dot (like :green_circle:) that indicates if you're currently active on Zulip, idle, offline, or unavailable.

Then it goes on to define the conditions for each of those "four possible availabilities". We find that a user's "availability" is defined by a combination of presence with whether they've manually "set [themself] as unavailable". The latter refers to the `away` property in `user_status`.

So with a boring status -- `away` false, represented in initial data by being absent -- their availability is determined purely by presence. That's exactly what we want to do in the absence of the user_status feature, too.

The same thing ends up holding for the other components of user status, namely the text and the emoji: when they haven't been set by the user, the thing we want to do is exactly the same as if the user status feature didn't exist.

That's what this comment was saying, though too tersely: the semantics of the user-status feature is such that when it's unset for a given user, the various things it would affect behave as if the feature didn't exist at all.

**Review by gnprice — COMMENTED** [review]
Thanks @chrisbobbe! Glad to see this coming along.

A couple of comments below just from reading the first three commits:

c5977ee35 types [nfc]: Remove unhelpful UserStatusMapObject type
ccb978d22 types: Stop sharing UserStatus with /register response type
c840447f3 types [nfc]: Remove misleading bit of jsdoc on initial-data `user_status`

Getting up now for dinner, but I had a long reply on one point, so I wanted to go ahead and post it for you :slightly_smiling_face:

**Review by gnprice — COMMENTED** [review]
OK, and comments below on reading most of the rest.

I think in particular I'd like to drop these commits:

3323c17c7 user status: Undo proliferation of unnecessary selectors (2/4)
1559a1788 user status: Undo proliferation of unnecessary selectors (3/4)

which take some of the logic of the model (the fact that an unexpressed `away` or `status_text` means false or empty-string, respectively) and move it out into callers. Instead, let's keep that inside the model.

It may still be useful to have their neighbors 1/4 and 4/4, or something like them, to consolidate the `getSelfUserStatus*` selectors. And that logic as to `status_text` should be applied uniformly; the old code is missing it in `getUserStatusTextForUser`.

**Review by chrisbobbe — COMMENTED** [review]
Thanks for the detailed review! 🙂 Revision pushed, and I've posted some replies to your comments.

File: src/user-status/userStatusReducer.js [hunk 3]
```diff
@@ -21,35 +54,69 @@ export default (
...
+        // The user had a boring status at /register time.
+        ?? kBoringUserStatus;
+
+      const newUserStatus = { ...oldUserStatus };
+      if (typeof action.away === 'boolean') {
+        newUserStatus.away = action.away;
+      }
+      if (typeof action.status_text === 'string') {
+        newUserStatus.status_text = action.status_text.length > 0 ? action.status_text : null;
+      }
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply src/user-status/userStatusReducer.js]: Oh! Somehow I missed this comment while working on the current revision. What do you think of this reasoning, quoting from my other comment https://github.com/zulip/zulip-mobile/pull/5224#discussion_r803162255:

> Incidentally, although either could work, I prefer `null` to `''`, mainly because Flow can't track separate types for empty and non-empty strings. Callers like the one in `UserStatusScreen` that want to use an empty string for their own UI choices, like clearing out an `<Input />`, can do so. But the model gives a nice and explicit `string` | `null`.

[the remark this answers is not in the document]
Comment (chrisbobbe) [reply src/user-status/userStatusReducer.js]: I've just pushed the simplification to use `action.away ?? oldUserStatus.away` but haven't yet touched `status_text`; let me know what you think. 🙂

File: src/api/modelTypes.js (hunk 1 above)
Comment (gnprice) [reply src/api/modelTypes.js:222]: Yeah, possibly. Are there other examples you have in mind, in addition to `Message`?

In the case of `Message`, the main difference within the API is on `flags`:
```js
  /**
   * The `flags` story is a bit complicated:
   *  * Absent in `event.message` for a `message` event... but instead
   *    (confusingly) there is an `event.flags`.
   *  * Present under `EVENT_NEW_MESSAGE`.
   *  * Present under `MESSAGE_FETCH_COMPLETE`, and in the server `/message`
   *    response that action is based on.
   *  * Absent in the Redux `state.messages`; we move the information to a
   *    separate subtree `state.flags`.
   */
  flags?: $ReadOnlyArray<string>,
```
where it feels to me like a mistake, not intentional. Though that doesn't mean we shouldn't distinguish the types; perhaps we should.

(In the case of `Message` I definitely wouldn't want to duplicate writing down all those other properties, though, because there are so many plus there's the subtleties in PMs vs. stream messages. We'd want to do something with spreads.)

File: src/api/initialDataTypes.js (hunk 2 above)
Comment (gnprice) [reply src/api/initialDataTypes.js:504]: > * Is "boring" a helpful label for anything here, or should we choose something else? In this draft I've got a `kBoringUserStatus`. Are you thinking that should change?

In a different API discussion recently I called something similar a "zero" value:
https://chat.zulip.org/#narrow/stream/378-api-design/topic/stream.20IDs.20in.20.60muted_topics.60/near/1325745
By that I mean the value that something has when there isn't any explicit record saying what value it has.

In that other context it's fundamental that there has to be a zero value, because that was about a relation between users and topics, and if we tried to explicitly store a value for every pair of a user and a topic that would be an enormous, quadratic amount of space. So we can't do that, so there must be many (user, topic) pairs with no explicit value, so there must be some interpretation of what that means, and that interpretation is the zero value.

Here, there didn't absolutely *have* to be a zero value -- it'd be perfectly feasible for every user to have an explicit status. But the design we chose does have a zero value. The API docs aren't real clear on this point, hmm -- particularly at `user_status` in https://zulip.com/api/register-queue -- but that does say:
> A dictionary which contains the status of all users in the Zulip organization who have set a status.

which means there can be users who have *not* "set a status", there won't be an explicit record for them, and there must be some interpretation of what that means. (And in [this chat message](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/Emoji.20statuses.20in.20zulip.2Eyaml/near/1322329) you laid out what the facts seem to be of that interpretation, which the docs should be amended to include.) So whatever the status is for users who have not "set a status" and don't appear in that data -- that's the zero status.

And that zero value is a bit more than just an arbitrary optimization in the data, too -- it reflects something real in the UX of this product feature. The zero user-status is the same as the user-status you have if you've just created your account and never interacted with the feature. I.e. it's the default, but moreover this isn't like language or something where there are many possible defaults which might vary between servers or realms -- there's really one canonical default that makes sense given the nature of this product feature. (I guess *maybe* there could be an org that wants a default `away` of true; but to have every user start with some status emoji, or some non-empty status text, would really not make sense unless it were an attempt to redefine the feature to serve some purpose it wasn't intended for.) That fact gives it a bit more weight and permanence in the API than if it were just an arbitrary optimization… though really even without that, so long as it is specified there in the API, that's enough already to make it legit for us to rely on it.

Anyway, if that discussion makes sense to you, then perhaps "zero" is a good term for more precisely getting at basically the same thing "boring" is expressing here, and it could be `kZeroUserStatus`.

Comment (gnprice) [reply src/api/initialDataTypes.js:504]: > However, I think it's possible that data is omitted in the first place (by status-aware servers) just because servers want to minimize how much data they send while still giving complete information.

Yeah, that is possible in principle, and there could be some feature in the future where there's a zero value in how the data's expressed (by servers that have the feature) and some different value that we want to use for handling a server that doesn't have the feature.

Here, though, those two values coincide. And while it doesn't have to *always* be that way, I think it's far from an accident that they do:
* When there's a zero value (in the sense of sometimes having no explicit record, and a value to interpret that as implicitly meaning), it's generally going to correspond to what happens when a user hasn't chosen to actively interact with something. After all, that's generally going to be the common case.
* Whatever happens when a user hasn't chosen to actively interact with something via some feature, that's generally going to be the same thing that happened with servers that didn't have the feature at all. That is, new features are much more often opt-in than opt-out.

In any case, it is true of *this* feature that the value to use in those two situations is the same; so that's something we can rely on.

And then it's not particularly odd that that's the case -- it's the way things can typically be expected to be. So it's good to have a brief comment noting it explicitly (to confirm that we did think about this question and check that this agreement holds for this feature), but then we can proceed to cheerfully and matter-of-factly rely on it.

Comment (gnprice) [reply src/api/initialDataTypes.js:504]: > We do have to choose whether or not to offer a UI to set these status components. We shouldn't show an emoji picker if the server doesn't support emoji statuses, and we shouldn't try to let people set themselves "unavailable" or choose a text status when those features aren't supported.
> 
> Happily, we have the feature level, so we don't have to go back to the edge to check for presence/shape of the status data we got straight from the server. slightly_smiling_face

Yep, agreed.

File: src/utils/presence.js [hunk 4]
```diff
@@ -114,7 +114,7 @@ export const statusFromPresence = (presence?: UserPresence): PresenceStatus => {
 };
 
 export const statusFromPresenceAndUserStatus = (
-  presence?: UserPresence,
-  userStatus?: UserStatus,
+  presence: UserPresence | void,
+  userStatus: UserStatus | void,
 ): PresenceStatus | 'unavailable' =>
```
[the remark this answers is not in the document]
Comment (gnprice) [reply src/utils/presence.js]: Yeah. I think really the ideal behavior here would be for Flow to reject a signature that claims to put an optional parameter before a required one -- it seems like there's no way such a type can end with people being happy.

File: src/utils/presence.js [hunk 5]
```diff
@@ -94,7 +94,7 @@ export const presenceToHumanTime = (presence: UserPresence, status?: UserStatus)
     : `${formatDistanceToNow(lastTimeActive)} ago`;
 };
 
-export const statusFromPresence = (presence?: UserPresence): PresenceStatus => {
```
Comment (gnprice) [inline src/utils/presence.js:97]: In commit message:

    See
      https://github.com/zulip/zulip-mobile/pull/5224/commits/aa40a6fff157fbc5e298369c09fd5529d727aa7b#r802272836

GitHub unfortunately has a habit of taunting you if you try to follow a link like this, one referring to both a PR and a commit ID:
image
> :telescope: 
> We went looking everywhere, but couldn’t find those commits.

(I say "taunting" partly because "went looking everywhere" is a bald-faced lie: if you edit the URL to just refer directly to the commit without the PR, GitHub finds the commit with no trouble. Like this: https://github.com/zulip/zulip-mobile/commit/aa40a6fff157fbc5e298369c09fd5529d727aa7b

That doesn't show the discussion *on* the PR, though, which is I think what you mainly meant to link to in this case.)

Instead, if you go to the main PR page and find the subthread there, you get a link like this: `https://github.com/zulip/zulip-mobile/pull/5224#discussion_r802272836` and those are more stable.

(There's another link like this one in a later commit too, "user status types: Say that /register response data is string-keyed".)

File: src/user-status/__tests__/userStatusReducer-test.js [hunk 6]
```diff
@@ -1,195 +1,180 @@
-import deepFreeze from 'deep-freeze';
+/* @flow strict-local */
```
[the remark this answers is not in the document]
Comment (gnprice) [reply src/user-status/__tests__/userStatusReducer-test.js:1]: > user status model: Use Immutable.Map<UserId, UserStatus> for Redux state

Yeah, this is the one I had in mind. (The other two make pretty small changes to that test file.)

Reading through the test changes in that commit now, they are all pretty straightforward. I'm fine with leaving these changes in the order they're in.

File: src/api/eventTypes.js [hunk 7]
```diff
@@ -100,23 +100,54 @@ export type PresenceEvent = $ReadOnly<{|
...
   status_text?: string,
+
+  // For these three emoji-status properties:
+  //
+  // non-empty string: User's status emoji is as indicated.
+  // "": User's status emoji is unset
+  // omitted: User did not change this; keep current value, or:
+  // TODO(server-5.0): Absent before feature level 86.
+  //
+  // "As indicated" means what the doc says, or:
```
Comment (gnprice) [inline src/api/eventTypes.js:135]: I don't quite understand this line. Is what follows meant to restate what's in the doc, or supplement it, or is it an alternative thing that the data could mean?

File: src/api/eventTypes.js [hunk 8]
```diff
@@ -100,23 +100,54 @@ export type PresenceEvent = $ReadOnly<{|
...
+ *   status
  *
- * @prop [away] - update user's away status:
- *       - `true` the user is away regardless of presence
- *       - `false` remove the away status, now use presence
- * @prop [status_text] - if present:
- *       - empty string clears the user's status text
- *       - any string sets user's status to that
+ * See InitialDataUserStatus for the corresponding initial data. The
+ * per-user properties have different meanings between here and there.
```
Comment (gnprice) [inline src/api/eventTypes.js:107]: Let's also link to the API docs:
  https://zulip.com/api/get-events#user_status

File: src/api/eventTypes.js [hunk 9]
```diff
@@ -100,23 +100,54 @@ export type PresenceEvent = $ReadOnly<{|
...
+  //
+  // Unicode emoji:
+  //   reaction_type is 'unicode_emoji'
+  //   emoji_code is the emoji's Unicode codepoint
+  // realm emoji:
+  //   reaction_type is 'realm_emoji'
+  //   emoji_code is the ID in the realm-emoji namespace
+  // Zulip extra emoji:
+  //   reaction_type is 'zulip_extra_emoji'
+  //   emoji_code is the ID in the Zulip extra emoji namespace
```
Comment (gnprice) [inline src/api/eventTypes.js:147]: I think this is all covered in the API docs (following the links from this event's doc to https://zulip.com/api/add-reaction#parameters ), so we don't need to repeat it here.

It would be good to include a pointer to the `Reaction` type, though, because that includes some information that isn't in the docs. And, hmm, one thing the docs say:
> (`emoji_code` will be its Unicode codepoint).

is not true, and is incomplete even in the cases where it's partly or arguably true.

The truth is that a Unicode emoji may consist of a sequence of several code points, and there's a particular idiosyncratic way they have to be encoded to come up with the Zulip `emoji_code`.

The jsdoc on our `ReactionType` accurately says that a Unicode emoji is a sequence of code points. The jsdoc on our `Reaction#emoji_code` adds a link to the implementation, where a comment (which I wrote, back in 2018, after reverse-engineering this) explains the actual encoding -- here's a direct link:
https://github.com/zulip/zulip/blob/73266350efa3b6a39dc36744e91782010f974481/zerver/models.py#L2903-L2907

At the time I wrote that, the API docs didn't exist. That information should probably make it back into the API docs now that they do exist, but in the meantime we can point at the existing accurate version in our comments.

**Comment by gnprice** [comment]
Aha. Just now I was looking at the reducer code in this PR, seeing about simplifying it. And I had a realization that I think can help simplify conceptually the picture of what's going on with this API.

The event and the initial data are really doing the same thing: a value is mentioned only when it's changed.

The point is that in the case of the initial data, "changed" means changed relative to the zero value: `{ away: false, status_text: '', emoji_name: '', reaction_type: '', emoji_code: '' }`.

In light of that point, I think it's also possible to re-unify the description of these properties into a single `UserStatus` type in `modelTypes.js`. Then where it appears in the initial data, the jsdoc just needs to make clear that it's relative to that zero value.

I think it should then also be possible to write the reducer's per-user update code once, and have both the REGISTER_COMPLETE and EVENT_USER_STATUS_UPDATE cases invoke that.

**Comment by gnprice** [comment]
> a single `UserStatus` type in `modelTypes.js`.

Or perhaps better named: `UserStatusUpdate`.

**Comment by gnprice** [comment]
Here's a version of what the reducer would then look like:
```js
type UserStatusUpdate = $ElementType<
  $NonMaybeType<$ElementType<InitialDataUserStatus, 'user_status'>>,
  string,
>;

function updateUserStatus(
  status: UserStatus,
  update: $ReadOnly<{ ...UserStatusUpdate, ... }>,
): UserStatus {
  const { away, status_text, emoji_name, emoji_code, reaction_type } = update;
  return {
    away: away ?? status.away,
    status_text: status_text ?? status.status_text,
    status_emoji:
      emoji_name != null && reaction_type != null && emoji_code != null
        ? reaction_type === ''
          ? null
          : { emoji_name, reaction_type, emoji_code }
        : status.status_emoji,
  };
}

// …

    case REGISTER_COMPLETE: {
      const { user_status } = action.data;
      if (!user_status) {
        // TODO(server-1.9.1): Drop this.
        return initialState;
      }

      return Immutable.Map(
        objectEntries(user_status).map(([id, update]) => [
          makeUserId(Number.parseInt(id, 10)),
          updateUserStatus(kBoringUserStatus, update),
        ]),
      );
    }

    case EVENT_USER_STATUS_UPDATE: {
      const oldUserStatus = state.get(action.user_id, kBoringUserStatus);
      return state.set(action.user_id, updateUserStatus(oldUserStatus, action));
    }
```

(The `type UserStatusUpdate = $ElementType<…>` is a quick hack -- in the actual code that type would be defined in `modelTypes.js`, and the initial data and event would both use that.)

**Comment by gnprice** [comment]
OK, done reading this revision. I think the high-order bit is in those top-level comments just now. Some background for those is in the subthread further above at https://github.com/zulip/zulip-mobile/pull/5224#discussion_r808451105 .

File: src/api/eventTypes.js (hunk 7 above)
Comment (chrisbobbe) [reply src/api/eventTypes.js:136]: I guess it was a too-terse way of saying, "If you're reading this after we've settled what actually belongs in the doc, then read the doc. Otherwise, read this, because we don't trust what the doc says now (e.g., it links to docs on reactions, which are unrelated)."

File: src/api/eventTypes.js (hunk 9 above)
Comment (chrisbobbe) [reply src/api/eventTypes.js:148]: I think actually the server doc should not link to any reaction-related docs. (I think it was a mistake to call one of these `reaction_type`.)

The right abstraction is "emoji", I think, not "reaction". It may be feasible for the server doc to deduplicate on an "emoji" concept—but I think it would be a fine improvement on the status quo for the doc to just spell out what is meant by each of these three "status emoji" properties.

Comment (gnprice) [reply src/api/eventTypes.js:148]: I agree that it's confusing to have "reaction" appear here, and is a sign that this area of the API isn't factored the right way.

I think it's not a coincidence that these three properties exactly coincide between status emoji and emoji reactions, though.

In reactions, the name `reaction_type`, and the use of the word "emoji" in each of the three existing values for that type, indicate that the idea when that was designed was that maybe there would be other kinds of "reactions" other than emoji, and the reactions would cover those. In that vision, it'd be a temporary coincidence that the structure of status emoji exactly coincides with the structure of reactions -- the wild and crazy unknown future concept of another "reaction" type that we might add might not make sense for user status.

But I think that early vision isn't happening. It's been years, and we have no feature ideas in which we'd want to overload "reactions" with something that isn't an emoji. If we did someday add a fourth value of `reaction_type`, I think it would basically still have to be some kind of emoji -- that's really what the UX of the reactions feature calls for. Something that truly doesn't fit as an emoji isn't going to fit in that reactions UX either, and so will use its own new bit of API surface.

That means that the structure we had for describing a "message reaction" is really all about an "emoji reaction" -- which means it's really just describing an emoji, plus saying what message to attach it to and on whose behalf. It just has this one quirk where one of the properties is named `reaction_type` and should really be `emoji_type`.

Given that, when there's this other feature that wants to describe an emoji, it's useful for the API to avoid multiplying concepts and reuse the "describe an emoji" concept it already has. Especially so given how gosh-darn *complicated* that concept is, as described at my `models.py` link above. That raises the cost (for the user of the API) of duplicating all that detail.

I think my ideal version of this status API would have called the property `emoji_type` instead of `reaction_type`. Then these three properties would exactly correspond to a generic "emoji" concept; and the doc on message reactions could point to here, or to a factored-out page elsewhere, and just mention that they have a quirk that they say `reaction_type` instead of `emoji_type`.

Comment (chrisbobbe) [reply src/api/eventTypes.js:148]: Ah OK, yeah, that's reasonable.

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed. Please feel free to review just the first _n_ commits, and I can split the rest off into another PR to make things more manageable. 🙂

One thing that I think remains: in our model, do we want to represent a null status text as `''` or `null`; I could go either way but mentioned some points at https://github.com/zulip/zulip-mobile/pull/5224#discussion_r804345618.

File: src/user-statuses/userStatusesReducer.js [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+};
+
+function updateUserStatus(
+  status: UserStatus,
+  update: $ReadOnly<{ ...UserStatusUpdate, ... }>,
+): UserStatus {
+  const { away, status_text, emoji_name, emoji_code, reaction_type } = update;
+  return {
+    away: away ?? status.away,
+    status_text: status_text ?? status.status_text,
```
Comment (gnprice) [inline src/user-statuses/userStatusesReducer.js:42]: I believe this means that if an update comes in with `status_text: ''`, meaning there is no status text set, we'll get `status_text: ''` in our internal state, whereas the value the UserStatus jsdoc says should represent that is `status_text: null`. IOW, this logic and that type definition disagree on whether to use `null` or `''`.

I agree with the reasoning at https://github.com/zulip/zulip-mobile/pull/5224#discussion_r804345618 that `null` is nice here because Flow can distinguish it from strings. To reflect that choice here, I think it just needs to get a bit more complicated in a similar way to `status_emoji` below:
```suggestion
    status_text: status_text != null ? status_text === '' ? null : status_text : status.status_text,
```

Comment (gnprice) [reply src/user-statuses/userStatusesReducer.js:42]: This is probably also a good prompt to add some more test cases, in particular covering both updating from the zero value to something else and vice versa.

Hmm, also covering status emoji.

File: src/user-statuses/__tests__/userStatusReducer-test.js [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    });
+
+    test('when the user_ status text is updated this is reflected in the state', () => {
+      expect(
+        userStatusesReducer(Immutable.Map(), {
+          id: 0,
+          type: EVENT_USER_STATUS_UPDATE,
+          user_id: makeUserId(1),
+          status_text: 'Hello, world!',
+        }),
```
Comment (gnprice) [inline src/user-statuses/__tests__/userStatusReducer-test.js:85]: One thing you could do to help make these test cases less tedious to write (to help with then writing some new ones) is a `mkAction`-type helper -- in fact probably just taking a `UserStatusUpdate` as argument, and always using `eg.selfUser`.

File: src/user-statuses/__tests__/userStatusReducer-test.js [hunk 12]
```diff
@@ -0,0 +1,117 @@
+/* @flow strict-local */
```
Comment (gnprice) [inline src/user-statuses/__tests__/userStatusReducer-test.js:1]: This file's basename (as well as the directory above it) should probably change as part of the rename:
`src/user-statuses/__tests__/userStatusReducer-test.js`

**Review by gnprice — COMMENTED** [review]
Thanks for the revision!

Just a few comments below. I'd also make some tweaks to the jsdoc and comments, but that's probably simplest if I just do them on top at the end of the branch.

The first 16/21 commits, everything before the main commit, is good to merge. But at this point I think in fact the whole thing is quite close to merge, so probably simplest is to keep it all on this PR rather than split the main commit to a separate one. That way the main commit goes in with this PR where the discussion on it happened. Let's definitely leave the UI for a separate PR, though :slightly_smiling_face:

File: src/user-status/userStatusReducer.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (gnprice) [reply src/user-status/userStatusReducer.js]: Yeah, I like that reasoning for `null`: that it lets Flow track the difference between that and a value that's set. Continued below: https://github.com/zulip/zulip-mobile/pull/5224#discussion_r813550160

File: src/user-statuses/__tests__/userStatusReducer-test.js (hunk 11 above)
Comment (chrisbobbe) [reply src/user-statuses/__tests__/userStatusReducer-test.js:85]: > and always using `eg.selfUser`

It often ends up being important to coordinate the IDs between the state objects (initial and expected) and the action. Maybe good to have a helper for constructing state objects? (And maybe that helper should use the reducer itself, like how we make `baseState` in `unreadModel-test.js`?)

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed, and I've left a question at https://github.com/zulip/zulip-mobile/pull/5224#discussion_r814359774.

**Comment by gnprice** [comment]
OK, merging! Thanks for all your work on this.

I've added a few comment-commits on top -- take a look. One is actually cherry-picked from my draft branch of a few weeks ago for acting on TODO-server comments / relying more on non-ancient servers.

File: src/user-statuses/__tests__/userStatusReducer-test.js (hunk 11 above)
Comment (gnprice) [reply src/user-statuses/__tests__/userStatusReducer-test.js]: Yeah, possibly. Though I think these tests are fine as they are, in the new revision -- so further abstraction like that would mainly be useful if we were adding further complexity to this model, calling for more tests.

One thing that's going on in unreadModel-test.js, particularly in the EVENT_UPDATE_MESSAGE and EVENT_UPDATE_MESSAGE_FLAGS sections, is that:
* there's one kind of update that's very straightforward to implement -- adding a new message with EVENT_NEW_MESSAGE -- and can be used to build up an arbitrary state;
* then there's some other kinds of updates -- those other two event types -- where the implementation is much more complex, requiring careful testing.

So that first, very straightforward type of event makes a great way to build up a test state for use in testing the other types.

This is also a bit like how in the message-list diffing tests we use the fact that diffing from empty to a given list, and applying such a diff, are much simpler than making and applying a diff between two arbitrary lists. So to check if we've correctly handled an arbitrary diff, we make and apply those two much simpler diffs to get the before and after pictures of the DOM.

Here in the user-status model, there isn't so much of a gap between different levels of complexity. The thing we do when an update comes in is rather more complex than the unread model's new-message case… and then the only other nontrivial thing we do is just that exact same logic in a loop.

Comment (chrisbobbe) [reply src/user-statuses/__tests__/userStatusReducer-test.js]: > Here in the user-status model, there isn't so much of a gap between different levels of complexity.

Interesting; thanks for this helpful reasoning!

**Comment by chrisbobbe** [comment]
Great, thanks for the careful review! I've read through those comment commits on top; thanks for those too! 🙂
